src/fabric/test/fabric2_update_docs_tests.erl

```diff
@@ -0,0 +1,222 @@
+    meck:new(erlfdb, [passthrough]),
+    Db.
+
+
+cleanup(#{} = Db) ->
+    meck:unload(),
+    ok = fabric2_db:delete(fabric2_db:name(Db), []).
+
+
+update_docs(Db) ->
+    ?assertEqual({ok, []}, fabric2_db:update_docs(Db, [])),
+
+    Doc1 = doc(),
+    Res1 = fabric2_db:update_docs(Db, [Doc1]),
+    ?assertMatch({ok, [_]}, Res1),
+    {ok, [Doc1Res]} = Res1,
+    ?assertMatch({ok, {1, <<_/binary>>}}, Doc1Res),
+    {ok, {1, Rev1}} = Doc1Res,
+    {ok, Doc1Open} = fabric2_db:open_doc(Db, Doc1#doc.id),
+    ?assertEqual(Doc1#doc{revs = {1, [Rev1]}}, Doc1Open),
+
+    Doc2 = doc(),
+    Doc3 = doc(),
+    Res2 = fabric2_db:update_docs(Db, [Doc2, Doc3]),
+    ?assertMatch({ok, [_, _]}, Res2),
+    {ok, [Doc2Res, Doc3Res]} = Res2,
+    ?assertMatch({ok, {1, <<_/binary>>}}, Doc2Res),
+    ?assertMatch({ok, {1, <<_/binary>>}}, Doc3Res).
+
+
+update_docs_replicated(Db) ->
+    Opts = [replicated_changes],
+
+    ?assertEqual({ok, []}, fabric2_db:update_docs(Db, [], Opts)),
+
+    Doc1 = doc(10, {1, [rev()]}),
+    ?assertMatch({ok, []}, fabric2_db:update_docs(Db, [Doc1], Opts)),
+    {ok, Doc1Open} = fabric2_db:open_doc(Db, Doc1#doc.id),
+    ?assertEqual(Doc1, Doc1Open),
+
+    Doc2 = doc(10, {1, [rev()]}),
+    Doc3 = doc(10, {1, [rev()]}),
+    ?assertMatch({ok, []}, fabric2_db:update_docs(Db, [Doc2, Doc3], Opts)),
+    {ok, Doc2Open} = fabric2_db:open_doc(Db, Doc2#doc.id),
+    ?assertEqual(Doc2, Doc2Open),
+    {ok, Doc3Open} = fabric2_db:open_doc(Db, Doc3#doc.id),
+    ?assertEqual(Doc3, Doc3Open).
+
+
+update_docs_batches(Db) ->
+    Opts = [{batch_size, 5000}],
+
+    Docs1 = [doc(9000), doc(9000)],
+
+    meck:reset(erlfdb),
+    ?assertMatch({ok, [_ | _]}, fabric2_db:update_docs(Db, Docs1, Opts)),
+    ?assertEqual(2, meck:num_calls(erlfdb, transactional, 2)),
+
+    lists:foreach(fun(#doc{} = Doc) ->
+        ?assertMatch({ok, #doc{}}, fabric2_db:open_doc(Db, Doc#doc.id))
+    end, Docs1),
+
+    Docs2 = [doc(10), doc(10), doc(9000), doc(10)],
+
+    meck:reset(erlfdb),
+    ?assertMatch({ok, [_ | _]}, fabric2_db:update_docs(Db, Docs2, Opts)),
+    ?assertEqual(2, meck:num_calls(erlfdb, transactional, 2)),
+
+    lists:foreach(fun(#doc{} = Doc) ->
+        ?assertMatch({ok, #doc{}}, fabric2_db:open_doc(Db, Doc#doc.id))
+    end, Docs2).
+
+
+update_docs_replicated_batches(Db) ->
+    Opts = [{batch_size, 5000}, replicated_changes],
+
+    Docs1 = [doc(Size, {1, [rev()]}) || Size <- [9000, 9000]],
+
+    meck:reset(erlfdb),
+    ?assertMatch({ok, []}, fabric2_db:update_docs(Db, Docs1, Opts)),
+    ?assertEqual(2, meck:num_calls(erlfdb, transactional, 2)),
+
+    lists:foreach(fun(#doc{} = Doc) ->
+        ?assertEqual({ok, Doc}, fabric2_db:open_doc(Db, Doc#doc.id))
+    end, Docs1),
+
+    Docs2 = [doc(Size, {1, [rev()]}) || Size <- [10, 10, 9000, 10]],
+
+    meck:reset(erlfdb),
+    ?assertMatch({ok, []}, fabric2_db:update_docs(Db, Docs2, Opts)),
+    ?assertEqual(2, meck:num_calls(erlfdb, transactional, 2)),
```

Comment:
Shouldn't this be three transactions? I'd think it'd be `[10, 10], [9000], [1]`

Reply:
It would be 2 since [10, 10, 9000] is the first, then it would commit then another [10] would be the last.

Reply:
I wouldn't have expected you'd be able to exceed 5000 as a limit (with the exception of a single doc transaction).

Reply:
Not a big deal either way, to be clear. This is fine.